improve ways to support multiple platforms/versions

richm opened the pull request:
Add a couple of different ways to support defining platform/version
specific variables, and performing platform/version tasks.

richm committed:
improve ways to support multiple platforms/versions

Add a couple of different ways to support defining platform/version
specific variables, and performing platform/version tasks.

richm commented:
@pcahyna @robled @greg-hellings ptal when convenient

odyssey4me commented on README.md:
```diff
@@ -197,25 +197,70 @@ need different templates for different distributions.
+        - "setup_{{ ansible_os_family }}.yml"
+      paths:
+        - "{{ role_path }}/tasks"
+      skip: true
```
Note that with_first_found is deprecated, and its replacement of using a lookup does not work with errors='ignore' for include_vars. See https://github.com/ansible/ansible/issues/58942 for more details.

richm replied:
> Note that with_first_found is deprecated, and its replacement of using a lookup does not work with errors='ignore' for include_vars. See [ansible/ansible#58942](https://github.com/ansible/ansible/issues/58942) for more details.

The use of `with_<lookup>` syntax is not deprecated - see https://docs.ansible.com/ansible/latest/user_guide/playbooks_loops.html#
"Note
...
    We have not deprecated the use of with_<lookup> - that syntax will still be valid for the foreseeable future."

If you have different information, please let me know.  If you are referring to the error message emitted by the loop when using `skip: true`, I see no such message when using my simple test on Ansible 2.9.9:
```yaml
- name: Set version-specific variables
  include_vars: "{{ item }}"
  with_first_found:
    - files:
        - no_such_file.yml
      paths: "{{ role_path }}/vars"
      skip: true
```
Using `ANSIBLE_STDOUT_CALLBACK=debug ansible-playbook -vv role-var-test.yml` 
I see this output
```
TASK [role_b : Set version-specific variables] *********************************
task path: /home/rmeggins/ansible_sandbox/roles/role_b/tasks/main.yml:10
...

```
(that is, there is no output)
and I see this:
```
PLAY RECAP *********************************************************************
localhost                  : ok=7    changed=0    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0
```
That is, the non existent file is silently skipped.

odyssey4me replied:
He, I see https://github.com/ansible/ansible/commit/707e33793d683254a511cd1ec825df86a5121feb - I wonder whether that got ported back to 2.8 and therefore which range of versions will show the deprecated message. How confusing. :)

richm replied:
> He, I see [ansible/ansible@707e337](https://github.com/ansible/ansible/commit/707e33793d683254a511cd1ec825df86a5121feb) - I wonder whether that got ported back to 2.8 and therefore which range of versions will show the deprecated message. How confusing. :)

Will use `lookup` based on guidance from Ansible team.

richm commented:
Also, since comments are closed on https://github.com/ansible/ansible/issues/58942 - I'd like to carry on the conversation here, and possibly close https://github.com/oasis-roles/meta_standards/pull/24 as a duplicate of this PR.

richm commented:
Here is an example of using skip with the lookup function:
```yaml
- name: Set version-specific variables
  include_vars: "{{ item }}"
  loop: "{{ lookup('first_found', myloopparam) }}"
  vars:
    myloopparam:
      files:
        - no_such_file.yml
      paths: "{{ role_path }}/vars"
      skip: true
```
I see the same output as with the `with_first_found` case - the missing file is silently (as in - no task output) skipped, with `skipped=1` reported in the final output.

richm commented:
I general, I prefer the `with_first_found` syntax over the lookup function - it just looks cleaner to me.  But if someone can confirm that `with_first_found` is indeed deprecated, then of course we should use the lookup function instead.

richm commented:
Also note that due to ansible scoping, the variable `myloopparam` is not made global, so naming doesn't matter with respect to "polluting" the global namespace, although we may still want to require the use of a `__rolename_` prefix for this varname to be consistent with conventions.

nhosoi commented:
Never mind. RHELPLAN-46757 explains.
~I'm confused at the `lookup` proposed in~ https://github.com/oasis-roles/meta_standards/pull/24.~ ~I simply replaced `with_first_found` with `lookup` `first_found`, then it fails with `No file was found` [2] unless adding `{{ role_path }}/vars/` in front as in [1]. (please note that my test case is not flat, but the task is nested.)~

~If I use `with_first_found`, the files are found without specifying the path, while it does not work with `lookup`. What could be the difference? Was I just lucky when I used `with_first_found`? Or `lookup` is more strict on the relative path?~

greg-hellings commented on README.md:
```diff
@@ -197,25 +197,70 @@ need different templates for different distributions.
+    - "{{ role_path }}/vars/{{ ansible_distribution }}.yml"
+    - "{{ role_path }}/vars/{{ ansible_distribution }}_{{ ansible_distribution_major_version }}.yml"
+    - "{{ role_path }}/vars/{{ ansible_distribution }}_{{ ansible_distribution_version }}.yml"
+  when: item is file
```
Oooo, I like this. Big :+1:

greg-hellings commented on README.md:
````diff
@@ -197,25 +197,70 @@ need different templates for different distributions.
+      paths:
+        - "{{ role_path }}/tasks"
 ```
+Then you would provide `tasks/setup_default.yml` to do the generic setup, and e.g. `tasks/setup_Fedora.yml` to do the Fedora specific setup.  The "setup_default.yml" is required in order to use `with_first_found`, which will give an error if no file is found.
````
This line is too long.

Also, why not suggest using a `setup/` directory instead of forcing tasks to be a flat list of files?

richm replied:
fixed in https://github.com/oasis-roles/meta_standards/pull/23/commits/d8266d8254ae62e3e22c4d813a0b5c44d2b6e781

greg-hellings commented on README.md:
````diff
@@ -197,25 +197,70 @@ need different templates for different distributions.
+```
+The advantage of this is that you do not have to provide a
+`setup-default.yml`.  The disadvantage is that if you misspell a filename e.g.
+`setup-Feodra.yml` you will not get an error.  **NOTE**: Always specify the
````
In these two lines you use "-", which matches current /actual/ practices. But it doesn't match all the standards above.

Which are we going with?

richm replied:
oops - should have used `_` instead of `-` :P

richm replied:
fixed in https://github.com/oasis-roles/meta_standards/pull/23/commits/d8266d8254ae62e3e22c4d813a0b5c44d2b6e781

richm committed:
address review issues

richm committed:
added a note about including relative files with variables in the name

richm committed:
address more review comments

Use `loop` with `lookup('first_found')` instead of `with_first_found`.
Always specify the explicit, absolute path to the files to be included,
using `{{ role_path }}/vars` or `{{ role_path }}/tasks`.
Use the `ansible_facts['name']` bracket notation rather than the
`ansible_facts.name` or `ansible_name` form.

richm commented:
Thanks for all of the reviews.  Please check the latest commits.

greg-hellings commented:
All of these `ansible['distribution']` and `ansible['os_family']` need to be `ansible_facts...`

richm committed:
use ansible_facts instead of ansible

richm commented:
> All of these `ansible['distribution']` and `ansible['os_family']` need to be `ansible_facts...`

fixed in https://github.com/oasis-roles/meta_standards/pull/23/commits/eac20691c7abb98e753fe78ef9613e0a574bce38

richm commented:
Please review. I'd like to finalize this proposal because it is blocking the work we need to do to change all of the system roles.
